kasparovsyncd/sync.go

```diff
@@ -235,238 +255,594 @@ func doesBlockExist(dbTx *gorm.DB, blockHash string) (bool, error) {
 	return !httpserverutils.IsDBRecordNotFoundError(dbErrors), nil
 }
 
-// addBlocks inserts all the data that could be gleaned out of the verbose
-// block and raw block data into the database. This includes transactions,
-// subnetworks, and addresses.
-// Note that if this function may take a nil dbTx, in which case it would start
-// a database transaction by itself and commit it before returning.
-func addBlock(client *jsonrpc.Client, rawBlock string, verboseBlock rpcmodel.GetBlockVerboseResult) error {
+func addBlocksAndTransactions(client *jsonrpc.Client, blocks []*rawAndVerboseBlock) error {
 	db, err := database.DB()
 	if err != nil {
 		return err
 	}
 	dbTx := db.Begin()
 	defer dbTx.RollbackUnlessCommitted()
 
-	// Skip this block if it already exists.
-	blockExists, err := doesBlockExist(dbTx, verboseBlock.Hash)
+	transactionIDtoTxWithMetaData, err := insertBlocksTransactions(dbTx, client, blocks)
 	if err != nil {
 		return err
 	}
-	if blockExists {
-		dbTx.Commit()
-		return nil
+
+	err = insertBlocksTransactionOutputs(dbTx, transactionIDtoTxWithMetaData)
+	if err != nil {
+		return err
 	}
 
-	dbBlock, err := insertBlock(dbTx, verboseBlock)
+	err = insertBlocksTransactionInputs(dbTx, transactionIDtoTxWithMetaData)
 	if err != nil {
 		return err
 	}
-	err = insertBlockParents(dbTx, verboseBlock, dbBlock)
+
+	blockHashToID, err := insertBlocks(dbTx, blocks, transactionIDtoTxWithMetaData)
 	if err != nil {
 		return err
 	}
-	err = insertRawBlockData(dbTx, rawBlock, dbBlock)
+
+	err = insertTransactionBlocks(dbTx, blocks, blockHashToID, transactionIDtoTxWithMetaData)
 	if err != nil {
 		return err
 	}
 
-	blockMass := uint64(0)
-	for i, transaction := range verboseBlock.RawTx {
-		dbSubnetwork, err := insertSubnetwork(dbTx, &transaction, client)
-		if err != nil {
-			return err
+	dbTx.Commit()
+	log.Infof("Added %d blocks", len(blocks))
+	return nil
+}
+
+func insertBlocks(dbTx *gorm.DB, blocks []*rawAndVerboseBlock, transactionIDtoTxWithMetaData map[string]*txWithMetaData) (map[string]uint64, error) {
```

Comment:
Please give names to outputs, as it's not easily clear what the function returns